Noisy network and quantum limits

rdireen opened the pull request:
### Purpose
My friend and I added the ability to combine scikit-rf Networks in a way that correctly handles noise covariance calculations. Meaning you can calculate noise parameters (nfmin, y_opt, nf, etc.) after combining Networks in various ways: cascade, parallel-parallel, series-series. 

In addition, the calculations handle quantum limits at very cold and/or high frequencies. This code is being used to analyze networks that are cryogenically cooled to temperatures near 0 K. With our additions you should be able to play with and design LNAs.

### Examples
You can see our examples under doc/examples/noisynetworks or you can compile the sphinx documentation and look under Examples/Network Noise Analysis. The best examples to start with are **"Noise Analysis of Common Emitter Amplifier with Two-Port Networks"** and **"Multiport Networks Exposed to Multiple Temperatures and Quantum Behavior."**

### ToDo
This is our initial pull request to see what you all think, but there are additions and improvements we would like to make. We have only started in on unit tests, so we haven't adequately covered the code. We also acknowledge that we need more documentation. But, we have added some examples and have done a first pass of API documenting. 

Please let me know what you think, we would love to collaborate with you on your excellent project! The e-mail associated with my GitHub account would work fine for any questions.

jhillairet commented:
Wow, it's a huge PR, thanks. 

Are all the modified files intentively? I mean, for example, there are a lot of notebooks modified, which seems out of the context of the PR purpose. Do you think you could clean the PR with only the necessary modifications?

jhillairet commented:
this will also interest @FournierVincent

rdireen commented:
Ah, yeah I meant to do that, I'll clean that up in the next few days and
submit another pull request.

Randy

On Wed, Oct 14, 2020 at 1:59 PM Julien Hillairet <notifications@github.com>
wrote:

> Wow, it's a huge PR, thanks.
>
> Are all the modified files intentively? I mean, for example, there are a
> lot of notebooks modified, which seems out of the context of the PR
> purpose. Do you think you could clean the PR with only the necessary
> modifications?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-rf/scikit-rf/pull/401#issuecomment-708628058>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6RIBWN7C7KJPPK6WP3PYTSKX7KDANCNFSM4SQHGCRQ>
> .
>


-- 
Randal Direen, Ph.D.
DireenTech Inc.
[+1](719)330-8997

rdireen commented:
Ok, I did some cleanup. There were several examples that we ran, got commit happy, and accidentally pushed into the branch. I reverted those changes to what they were before.

But, you will notice a bunch of *.snp files that have been changed. While developing on a couple different OSs I noticed inconsistencies between files and their line endings. Some *.snp files had CRLFs at the end of lines while others just had LFs. I fixed this by adding a .gitattributes file in the main folder (_see_ [gitattributes](https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings)). The result of doing this meant that many *.snp files needed to be filtered so that they all have the same line ending, and those are the ones that you see in this pull request. The line endings are the only changes we made to those files, we didn't modify them in any other way.

Let me know if you see anything else that seems suspicious, I believe I caught everything.

While I'm talking too much anyway. If you like our changes, and want to integrate what we have, I should draw your attention to some overlapping capabilities between our code sets:

1. We have an object called MultiNetworkSystem that is very much like your circuit.py code, and
2. We added a _components_ directory that contains circuit elements. 

There are two major differences between our implementations and yours: first, our code handles noise calculations and transforms while yours doesn't; second, your code correctly handles complex reference impedances, while ours assumes Z0 is real. 

If your interested down the road, we can talk about tying the code together into something more cohesive. I didn't want to cut into your code too deeply before asking. You'll notice that most of my invasive additions have been within network.py

73, 
Randy

arsenovic commented:
randy, this is great. 

question: would it make sense to have a subclass of network for noise parameters, like a NoisyNetwork?     this would allow us to keep all noise-related stuff in a seperate file,  keep Network robust, and  make working on both easier.  the disadvantage is that you always have to explicitly use NoisyNetwork ?
 
what do you think? 

eventually removing over lap of the other components would be good also.

rdireen commented:
Hi Alex,

I do like the idea of breaking things into smaller modules. Also, explicitly defining a network as a NoisyNetwork is a nice way for the user to express their intent that they want to do a noise calculation. 

Something that would make defining a NoisyNetwork subclass a little sticky is if we want to simulate other network metrics. For example, we could associate an IP3 value with each Network and calculate a total IP3 value after cascading multiple Networks. Would this calculation be performed in Network, NoisyNetwork, or would we create an IP3Network subclass? If we create an IP3Network subclass and want to calculate both noise and IP3 values for cascaded networks, we would essentially need to do cascade operations twice (one for the set of NoisyNetworks and one for the set of IP3Networks).

Here is an option I thought of today: what if we create a subclass of Network called a Device, Component, or Ntwk? For example,


    class Ntwk(Network):

        __init__(self, args):

           self.noise_covariance = NetworkNoiseCov()

           self.ip3 = IP3()

So a Ntwk would be a Network that has a noise covariance matrix, ip3 values, etc. Cascading and combining Ntwks could handle all the additional calculations for noise and other metrics of interest. Then the Ntwk code would be separate from Network, keeping network.py smaller and independent of stuff you develop.  

Would this be a helpful way of going about things?

arsenovic commented:
yeah i see your point.  i dont think having two network classes  would be useful, and having infinite specialties;  NoisyNetwork , IP3Network, MixedMode, etc is not good either. each of these are special-ish cases   that make the code harder to read/learn, reduce robustness, etc.  but when they are 'just there' its very easy and nice for the user. 

i think the NoisyNetwork is different in that it is has larger scope,   a lot of the code will be run all the time, and it represents different physics. this  could justify a NoisyNetwork vs the other things, which are more or less just derived calcs on s-params. 

just thinking outloud.  @jhillairet  what are your thoughts?

jhillairet commented:
I don't know much to noise things, so I would be glad to learn more. @rdireen, I would be glad if you could indicate me some resources.

> So a Ntwk would be a Network that has a noise covariance matrix, ip3 values, etc. Cascading and combining Ntwks could handle all the additional calculations for noise and other metrics of interest. Then the Ntwk code would be separate from Network, keeping network.py smaller and independent of stuff you develop.

To my understanding, this is exactly what Alex was meaning with a `NoisyNetwork` class. Such class could: 
- inherit from `Network`, so embark all `Network` parameters and methods ;  
- adds the noise figure of merits and noise calculation methods ;
- allows operation between NoisyNetworks. 
To solve the problem you were referring to, if classic `Network`(s) are part of operations (cascading, etc) with `NoisyNetwork`(s), these `Network`(s) should be converted to `NoisyNetwork`(s).

mgrando commented:
Hey, My name is Maurio. I worked with Randy on this code. I think the best reference for the S-parameter noise covariance is listed in the ipython notebook "Multiport Networks Exposed to Multiple Temperatures and Quantum Behavior" the papers title is "Noise in Lossless Microwave Multiports" by Quiterio Garcia-Garcia [García-García - 2004 - Noise in lossless microwave multiports Noise in L.pdf](https://github.com/scikit-rf/scikit-rf/files/5409593/Garcia-Garcia.-.2004.-.Noise.in.lossless.microwave.multiports.Noise.in.L.pdf)
.  Its really easy to follow and works out a lot of examples. 

To cascade noise parameters you will always need the S-parameters as well. You also need S-parameters to derive the noise figure and noise temperature of the system so both modules need to be linked somehow or there will be a lot of redundant calculations.  I also like having a user explicitly call on the noise network stuff when needed. 

To get a feel for the noise cascade process you should check out "Wave Techniques for Noise Modeling and Measurement" by -Scott W. Wedge [Wedge and Rutledge - 1992 - Wave techniques for noise modeling and measurement.pdf](https://github.com/scikit-rf/scikit-rf/files/5409589/Wedge.and.Rutledge.-.1992.-.Wave.techniques.for.noise.modeling.and.measurement.pdf).  We have a way smoother approach then Wedge but yield the same results.

rdireen commented:
@jhillairet and @arsenovic, I did some thinking over weekend and I think I like the idea of putting the code into NoisyNetwork. I like that it makes the code more modular and I like that it makes the user have to explicitly decide that they are working with NoisyNetworks. 

If this sounds good to you guys I'll work on the code this weekend and try to have something out next week using the NoisyNetwork subclass of Network.

Does this sound good @mgrando?

mgrando commented:
You bet it does @rdireen ! I will be available this weekend to help as well. I need to change a figure in the ipython notebook in the temperature with quantum limits section. It was taken from the ADMX collaboration I work with. I want to replace it with a generic one so I don't get in trouble.  👍

arsenovic commented:
this sounds great to me.  look forward to it .

rdireen commented:
@jhillairet and @arsenovic, it took me a little while because of some other projects I have around here, but I split off our code into a noisyNetwork.py module. I also moved our noisy circuit components into a noisyComponents folder and moved our multiNetworkSystem.py module to multiNoisyNetworkSystem.py. 

I verified that all the unit tests still pass, and all our examples have been updated to work with the noisyNetwork.py module. 

I also verified that the sphynx documentation builds and that our documentation looks reasonable. I'm not as familiar with sphynx as I would like to be, so I don't know what warnings are reasonable.

@mgrando take a look at your examples and make sure I didn't mess them up. I had to add functionality that would take Network objects and make NoisyNetwork objects out of them.

Remember that several of the text files will show up as being changed because I added the .gitattributes file (see comment above). And, there are also some changes and updates I made to network.py that should be reviewed to make sure I'm not stepping on any toes.

Let me know what you think, I have some time coming up this week and the next that I can devote to the project.

jhillairet commented:
Thank you for this important work. Please give us some time to review it ;)

jhillairet commented on .gitattributes:
```diff
@@ -0,0 +1,12 @@
+# Set the default behavior for line endings, in case people don't have core.autocrlf set.
```
I recommend removing the vscode-induced changes and files: .gitattributes and .gitignore

rdireen replied:
removed .gitattributes and changes made within .gitignore.

jhillairet commented on doc/source/examples/metrology/three_receiver_cal/data/simulation.s2p:
```diff
@@ -1,1618 +1,1618 @@
-! Gamma         !               0.280297050603356 1590.92705735488 0.28029604186875 1590.92899160015 
-! Port Impedance275.336881934298 0.097737745415425 275.358754002361 0.0979074110842495 
-
+! Touchstone file from project Project26
```
Would be better to revert unmodified files to their original state to avoid clutter in the PR

rdireen replied:
Reverted all files that were only different in how they handled line feeds (i.e. CRLF vs LF)

jhillairet commented on doc/source/tutorials/data/ntwk1.s2p:
```diff
@@ -1,4 +1,4 @@
-! Created Thu Nov 11 11:09:06 2010
+! Created Thu Nov 11 11:09:06 2010
```
same comment, would be nice to remove unchanged files from the PR

rdireen replied:
Reverted all files that were only different in how they handled line feeds (i.e. CRLF vs LF)

jhillairet commented on skrf/network.py:
```diff
@@ -467,6 +465,68 @@ def from_z(cls, z, *args, **kw):
         me.z = z
         return me
 
+    @classmethod
```
Usefull additions

rdireen replied:
Thanks!

jhillairet commented on skrf/network.py:
```diff
@@ -4523,6 +4588,8 @@ def s2z(s, z0=50, s_def=S_DEF_DEFAULT):
         # Power-waves. Eq.(19) from [3]
         # Creating diagonal matrices of shape (nports,nports) for each nfreqs
         F, G = npy.zeros_like(s), npy.zeros_like(s)
+        F = F.astype(dtype=npy.complex)
+        G = G.astype(dtype=npy.complex)
```
this generates a copy of F, G. Did you run into troubles with the initial type? If yes, to avoid an additional copy, maybe add the dtype= parameter in the zeros_like ?

rdireen replied:
I did run into trouble. I'm not sure why I didn't just put these type definitions in the zeros_like to begin with. It's fixed now.

jhillairet commented on skrf/network.py:
```diff
@@ -4624,6 +4691,8 @@ def s2y(s, z0=50, s_def=S_DEF_DEFAULT):
         # Power-waves. Inverse of Eq.(19) from [3]
         # Creating diagonal matrices of shape (nports,nports) for each nfreqs 
         F, G = npy.zeros_like(s), npy.zeros_like(s)
+        F = F.astype(dtype=npy.complex)
+        G = G.astype(dtype=npy.complex)
```
same comment

rdireen replied:
Fixed

jhillairet commented on skrf/network.py:
```diff
@@ -5168,6 +5257,8 @@ def y2s(y, z0=50, s_def=S_DEF_DEFAULT):
     if s_def == 'power':
         # Creating diagonal matrices of shape (nports,nports) for each nfreqs 
         F, G = npy.zeros_like(y), npy.zeros_like(y)
+        F = F.astype(dtype=npy.complex)
+        G = G.astype(dtype=npy.complex)
```
idem

rdireen replied:
Fixed

jhillairet commented on skrf/network.py:
```diff
@@ -5446,7 +5539,13 @@ def t2y(t):
+    return s2y(s, z0, s_def)
+
+def t2a(t, z0=50):
+
```
please add the docstring

rdireen replied:
Added docstrig

jhillairet commented on skrf/noisyComponents/attenuator.py:
```diff
@@ -0,0 +1,27 @@
+    def __init__(self, attenuation_db, name=None, comments=None, f_unit=None, T0=None, s_def='power', **kwargs):
+        super().__init__(name=name, comments=comments, f_unit=f_unit, s_def=s_def, **kwargs)
+
+        if attenuation_db < 0:
+            raise ValueError("attenuation_db must be a positive value")
+        
+        self._attenuation = 10**(-attenuation_db/20)
+
+        ovec = npy.ones(len(self.frequency))
+        zvec = npy.zeros(len(self.frequency))
+
+        attn_vec = network_array([[zvec,                   self._attenuation*ovec],
+                                     [self._attenuation*ovec,                    zvec]])
+
+        self.s = attn_vec
```
Similar in principle to the `media.attenuator`, for example it could be implemented as:

`self.s = rf.DefinedGammaZ0(frequency=frequency).attenuator(-attenuation_db).s`

jhillairet commented on skrf/noisyComponents/__init__.py:
```diff
@@ -0,0 +1,9 @@
+
+
+
+from .attenuator import *
+from .circulator import *
+from .rlc_series_2port import *
+from .rlc_shunt_2port import *
+from .bjt_ce_2port import *
+from .wilkinson_nport import *
```
I would recommand to postpone the `noisyComponents ` in a future PR, the time for `noisyNetwork `API to be integrated. 

On the API point of view, we need to discuss (@arsenovic ?). I see two things:
- does it have sense to have a bunch of new classe names, but knowing that these classes are in fact all `Network ` or `NoisyNetwork `objects. 
- If there is an interest in doing so, we could be more ambitious: Instead of a directory `noisyComponents/`, maybe it is worth creating a `components/` directory, with some noisy components inside (or as a subdir)

rdireen replied:
What I can do (at least temporarily) is place noisyComponents within the doc/examples/noisynetworks. That way all our examples will work, but we don't have to make noisyComponents a part of the core scikit-rf code. Does this sound good to you?

As far as adding new class names, I believe this is within the spirit of object oriented programing. It is true that all of these new classes are Networks or NoisyNetworks, in the same way that a dog or a cat is an animal. For example, you might make an animal class have 4 legs, a mouth and ears, but the derived dog class might add the ability to bark and the derived cat class have whiskers. An attenuator and a circulator are both Networks, so the natural object oriented mechanism for creating these devices in code is to derived them from Network using inheritance. 

However, that's more philosophy than anything. You could of course just create a Network and name it an attenuator.  Since these decisions are more up to you and the other creators of scikit-rf, I will happily defer to your judgement.

As far as an opinion goes, I think it would be nice to have a folder called components where all objects are derived from NoisyNetwork. Since NoisyNetwork just extends Network, all the functionality that you would expect from Network would be transferred to the derived component.

jhillairet commented on skrf/noisyNetwork.py:
```diff
@@ -0,0 +1,812 @@
+    =====================  =============================================
+    Operator               Function
+    =====================  =============================================
+    \+                     combines noisy networks in series
+    \|                     combines noisy networks in parallel
```
these operators have a different meaning in `Network`. I'm afraid this be a source of future confusion and issues

rdireen replied:
I agree. Would you be willing to permit adding a global flag that would enable changing the functionality? I did this before and it worked well. Basically you do something like rf.enable_new_operators() at the top of a script and it changes the functionality of the operators. What I like about using operators with two-port networks is that you can create an entire circuit and the result just looks like a line of algebra. This is far more convenient that having to call individual functions for each combination. 

But, again the ball is in your court. If you think this is too awkward I can be convince to drop this functionality. Or perhaps use different operators (if there are any left :D).

jhillairet replied:
It's an interesting feature, Is it working for Nport as well ?

jhillairet commented on skrf/multiNoisyNetworkSystem.py:
```diff
@@ -0,0 +1,421 @@
+   
+    """
+
+    def __init__(self):
```
passing a list of `NoisyNetwork ` to the constructor would be convenient.

jhillairet commented on skrf/multiNoisyNetworkSystem.py:
```diff
@@ -0,0 +1,421 @@
+        
+        return enumb, errors
+
+    def reduce(self):
```
I didn't go into the implementation, but isn't it equivalent to cascade multiple NoisyNetwork together :

`reduced_nntwk = nntwk1 ** nntwk2 ** ... ** nntwkN`

?

rdireen replied:
No, the point of multiNoisyNetworkSystem.py is to work with multiport NoisyNetworks. For example, you can combine multiport splitters and three port circulators with this object in any combination. The reduce() function returns a new NoisyNetwork that is not necessarily a two-port. 

This IS very similar to your circuit.py module. The problem is that I don't currently know how to combine noise covariance matrices in a way that works with your circuit.py code. But, one of the things @mgrando needs in his work is the ability to combine multiport noisy networks. So I was kinda stuck and needed a mechanism that would do something similar to circuit.py but work with NoisyNetworks. I'm open to suggestions if you have any ideas on how to perhaps make this a part of circuit.py

jhillairet replied:
ok, see my comment below

jhillairet commented on skrf/multiNoisyNetworkSystem.py:
```diff
@@ -0,0 +1,421 @@
+from .noisyNetwork import NoisyNetwork
+from .networkNoiseCov import NetworkNoiseCov
+
+class MultiNoisyNetworkSystem(object):
```
Before adding a new class to the skrf API (which will be a future source of issues ;), I'm asking myself "it is necessary?" 

If I understand well, a `MultiNoisyNetworkSystem ` , If only made of 2ports, consists in a multiple cascaded NoisyNetworks, am I right ? 
What if some NoisyNetwork in the list have more ports than 2 ? 
It sounds like a kind of `Circuit` for NoisyNetworks

rdireen replied:
Yeah, see my above comment. MultiNoisyNetworkSystem works with multi-port NoisyNetworks. So more ports than two-ports. I made a lousy example that only worked with two-ports, I should have made something with splitters and circulators or somethings. 

This is very much like Circuit for NoisyNetworks.

jhillairet replied:
OK, I understand. 

For the sake of clarity, why not calling it `NoisyCircuit` then ? OK, the way of defining such circuit is a bit different from Circuit's way, but the purpose looks like identical.

From the theoretical point of, it would be interesting to see if the Circuit's building algorithm taken from (Hallbjörner, P., 2003, . Microw. Opt. Technol. Lett. 38, 99–102. https://doi.org/10/d27t7m ) could be extended to noisy Network as defined in Garcia's 2004 paper.

jhillairet reviewed (commented):
My general comment is that there is a big work here so first of all, thank you very much for that. 

To the form, since most of the files of the PR haven't been changed (whitespace / end of lines), I would recommend to not include them in the PR and to let them unchanged. 

I would also suggest delaying the addition of the `noisyComponents/*.py `for a future PR : we must first address the `NoisyNetwork `and its associated `networkNoiseCov` to get consistent with Network API. (and logical for users). For example,  if the operators + and / have a different meaning between `Network` and `NoisyNetwork`, this may lead to future issues.

Please find my comments attached and don't forget this is only my point of view and can be of course discussed.

rdireen commented:
@jhillairet thank you very much for spending the time to review these potential additions. I can't get to the code today but will try to put some time in this week. I've read your comments and I am in general agreement with you. And of course you know skrf better than I so I greatly value your comments.

I will hopefully be providing you with updates soon.

rdireen commented on skrf/network.py:
```diff
@@ -397,8 +394,9 @@ def __init__(self, file=None, name=None, comments=None, f_unit=None, s_def=S_DEF
         self.port_names = None
 
         self.deembed = None
-        self.noise = None
+        self.noise = None # This is the ABCD form of the covariance matrix
         self.noise_freq = None
+        self.T0 = T0 # Temperature at measurement
 
```
I think that is a good idea, but I didn't want to step on the work @FournierVincent seems to have done. It looks like his work will pull in noise parameters from touchstone files. I'd be happy to move some of this over to the NoisyNetwork object if it doesn't break anything he was working on.

rdireen commented on skrf/network.py:
```diff
@@ -1073,7 +1137,8 @@ def a(self):
     @a.setter
     def a(self, value):
-        self._s = a2s(value, self.z0)
+        z = a2z(value)
+        self._s = z2s(z, self.z0)
```
It looks like a2z was fixed recently in commit a7619e926decef5e5d5dc66c34d7bb54d32afd8a . I've restored this property to its original form now.

rdireen commented on skrf/network.py:
```diff
@@ -1246,9 +1311,9 @@ def noisy(self):
       whether this network has noise
       """
       try:
-        return self.noise is not None and self.noise_freq is not None
+          return self.noise is not None and self.noise_freq is not None
       except:
-        return False
+          return False
 
```
Yeah I definitely agree. Maybe @FournierVincent can chime in to see what he thinks. I think this would require moving some of the file import functionality over to NoisyNetwork. This would allow touchstone files with noise parameters to be imported and worked with.

rdireen commented on skrf/network.py:
```diff
@@ -1267,16 +1332,13 @@ def n(self):
+          raise ValueError('network does not have noise')
       return self.noise_freq
 
     @property
```
agreed

rdireen commented on skrf/network.py:
```diff
@@ -1715,16 +1783,13 @@ def set_noise_a(self, noise_freq=None, nfmin_db=0, gamma_opt=0, rn=1 ) :
               )
           self.noise = noise.swapaxes(0, 2).swapaxes(1, 2)
           self.noise_freq = noise_freq
         
```
agreed

rdireen commented on skrf/vi/sa.py:
```diff
@@ -1,175 +1,175 @@
-        '''
-        self.write('rcls %i'%reg_n)
-
+
```
I restored this to its original with the CRLF appended to each line.

rdireen commented on skrf/util.py:
```diff
@@ -46,6 +46,19 @@
 except NameError:
     basestring = (str, bytes)
 
+def network_array(v):
+    '''
+    Convenience function for creating properly formed network matrix vectors
+
+     Examples
+    -----------
+    >>> rseries = npy.ones(len(frequency))
+    >>> r_series_y = skrf.network_array([[1/rseries,  -1/rseries], [-1/rseries,  1/rseries]])
+    >>> n_rseries = skrf.Network.from_y(r_series_y)
+
+    '''
+    v = npy.array(v)
+    return v.swapaxes(0, 2).swapaxes(1, 2)
```
I do agree that the name is bad, so that would need to change. 

My argument for keeping something like it is that it allows a user to input arbitrary networks. For example, suppose someone analytically determines  an admittance (Y) matrix for a network of resistors and capacitors, a utility like network_array is nice to have to generate the vector of matrices required by from_y. 

However, it's really not too hard to call val.swapaxes(0, 2).swapaxes(1, 2) on a numpy array so if you would prefer I can change my examples to do just that.  

My suggests would be to either:

1. rename network_array to something else (maybe _axes_swap_? )
2. use val.swapaxes(0, 2).swapaxes(1, 2) in the examples (probably introduces the least amount of clutter)

rdireen commented:
@jhillairet I've cleaned up the PR and addressed many of your comments. I'll get back to working on the rest of them this week. Thanks again for taking the time to consider our additions.

jhillairet commented:
> 
> 
> @jhillairet I've cleaned up the PR and addressed many of your comments. I'll get back to working on the rest of them this week. Thanks again for taking the time to consider our additions.

thanks, I'll check it. We can also wait for other's advices, which maybe differs from mine.

arsenovic commented:
if all the new additions are modular, could this be re-submitted easily as a fresh PR? just for the sake of readability?

rdireen commented:
@jhillairet and @arsenovic. Sorry for the delay, looks like December is going to be buys for me. After the new year I'll get back to work on this stuff. I'll fix some more of the pieces @jhillairet suggested and start a new pull request.

jhillairet commented:
Hi @rdireen 

Do you plan to continue your work on this? There is definitely a need for dealing with noisy networks (cf #538 for example)

jhillairet commented:
Update : the work made in this branch by @rdireen and @mgrando has been used in the following paper https://arxiv.org/pdf/2209.04008.pdf
